src/reuse/lint.py

```diff
@@ -8,329 +8,260 @@
-            out.write("\n")
-        out.write("\n")
-
-    return both | only_copyright | only_licensing
-
-
-def lint_summary(report: ProjectReport, out=sys.stdout) -> None:
-    """Print a summary for linting."""
-    # pylint: disable=too-many-statements
-    out.write("# ")
-    out.write(_("SUMMARY"))
-    out.write("\n\n")
-
-    file_total = len(report.file_reports)
-
-    out.write("* ")
-    out.write(_("Bad licenses:"))
-    for i, lic in enumerate(sorted(report.bad_licenses)):
-        if i:
-            out.write(",")
-        out.write(" ")
-        out.write(lic)
-    out.write("\n")
-
-    out.write("* ")
-    out.write(_("Deprecated licenses:"))
-    for i, lic in enumerate(sorted(report.deprecated_licenses)):
-        if i:
-            out.write(",")
-        out.write(" ")
-        out.write(lic)
-    out.write("\n")
-
-    out.write("* ")
-    out.write(_("Licenses without file extension:"))
-    for i, lic in enumerate(sorted(report.licenses_without_extension)):
-        if i:
-            out.write(",")
-        out.write(" ")
-        out.write(lic)
-    out.write("\n")
-
-    out.write("* ")
-    out.write(_("Missing licenses:"))
-    for i, lic in enumerate(sorted(report.missing_licenses)):
-        if i:
-            out.write(",")
-        out.write(" ")
-        out.write(lic)
-    out.write("\n")
-
-    out.write("* ")
-    out.write(_("Unused licenses:"))
-    for i, lic in enumerate(sorted(report.unused_licenses)):
-        if i:
-            out.write(",")
-        out.write(" ")
-        out.write(lic)
-    out.write("\n")
-
-    out.write("* ")
-    out.write(_("Used licenses:"))
-    for i, lic in enumerate(sorted(report.used_licenses)):
-        if i:
-            out.write(",")
-        out.write(" ")
-        out.write(lic)
-    out.write("\n")
-
-    out.write("* ")
-    out.write(_("Read errors: {count}").format(count=len(report.read_errors)))
-    out.write("\n")
-
-    out.write("* ")
-    out.write(
-        _("Files with copyright information: {count} / {total}").format(
-            count=file_total - len(report.files_without_copyright),
-            total=file_total,
-        )
-    )
-    out.write("\n")
 
-    out.write("* ")
-    out.write(
-        _("Files with license information: {count} / {total}").format(
-            count=file_total - len(report.files_without_licenses),
-            total=file_total,
+    def custom_serializer(obj):
+        """Custom serializer for the dictionary output of ProjectReport
+
+        :param obj: Object to be serialized
+        """
+        if isinstance(obj, PosixPath):
+            return str(obj)
+        if isinstance(obj, set):
+            return list(obj)
+        raise TypeError(
+            f"Object of type {obj.__class__.__name__} is not JSON serializable"
         )
+
+    return json.dumps(
+        report.to_dict(),
+        indent=2,
+        # Serialize sets to lists
+        default=custom_serializer,
+        sort_keys=True,
```

Comment:
Are we sure we want to sort keys? This means that the `json_version` and `reuse_version` keys are no longer at the top (assuming that `json.dumps` respects Python's insertion-ordered dictionaries).

Reply:
I personally wouldn't care about alnum sorting but about the fact that two runs of `reuse lint --json` return the very same output. Especially in larger projects I found it confusing in my tests that the order of keys and the list values aren't static.

If it were possible to sort all keys (and lists) but still have the version keys on top, that would be ideal. If not, I'd prefer the overall sorting to having the version on top.

Reply:
In the interest of getting this PR merged sooner rather than later, would it be OK for you two to create a new issue for this and then tackle it independently?

Reply:
As said, I would consider unsorted keys on all levels as a bug and therefore important to be fixed in this PR. But yes, we could split off the features of having the `version` keys on the top of the output IMHO.

Reply:
Just added a commit that should solve this.